File: frontend/src/types/execution.ts
export interface ExecutionStep {
  step_id: number;
  line_number: number | null;
  event_type: string;
  variables: Record<string, any>;
  stdout: string[];
  call_stack: any[];
  memory: Record<string, any>;
  visualization_events: any[];
  timestamp?: number;
}

File: backend/app/schemas/execution.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel
from .events import VisualizationEvent

class ExecutionStep(BaseModel):
    step_id: int
    line_number: Optional[int]
    event_type: str
    
    variables: Dict[str, Any]
    stdout: List[str]
    call_stack: List[Dict[str, Any]]
    memory: Dict[str, Any]
    visualization_events: List[VisualizationEvent]
    timestamp: Optional[float] = None

class ExecutionRequest(BaseModel):
    language: str = "python"
    code: str

class ExecutionResponse(BaseModel):
    execution_id: str
    status: str
    steps: List[ExecutionStep]
    error: Optional[Dict[str, Any]] = None
